### ISSUE TITLE
Tons of errored transfers, "database is locked" in logs

### ISSUE THREAD

**Issue by tylerball** [slskd/slskd#748]
Hi JP,

On latest pull from `canary` enqueuing downloads and uploads results in a lot of errors. Looking at the logs from docker-compose there are a ton of "database is locked" messages. For example:

```
slskd                   | [05:51:53 INF] Download of ***********.flac from user****** failed
slskd                   | [05:51:53 ERR] An exception occurred in the database while saving changes for context type 'slskd.Transfers.TransfersDbContext'.
slskd                   | Microsoft.Data.Sqlite.SqliteException (0x80004005): SQLite Error 6: 'database table is locked'.
slskd                   |    at Microsoft.Data.Sqlite.SqliteException.ThrowExceptionForRC(Int32 rc, sqlite3 db)                                                                                             slskd                   |    at Microsoft.Data.Sqlite.SqliteDataReader.NextResult()
slskd                   |    at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader(CommandBehavior behavior)                                                                                                 slskd                   |    at Microsoft.Data.Sqlite.SqliteCommand.ExecuteNonQuery()
slskd                   |    at Microsoft.Data.Sqlite.SqliteConnectionExtensions.ExecuteNonQuery(SqliteConnection connection, String commandText, SqliteParameter[] parameters)
slskd                   |    at Microsoft.Data.Sqlite.SqliteTransaction..ctor(SqliteConnection connection, IsolationLevel isolationLevel, Boolean deferred)
slskd                   |    at Microsoft.Data.Sqlite.SqliteConnection.BeginTransaction(IsolationLevel isolationLevel, Boolean deferred)
slskd                   |    at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.BeginTransaction(IsolationLevel isolationLevel)
slskd                   |    at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.Execute(IEnumerable`1 commandBatches, IRelationalConnection connection)
slskd                   |    at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(IList`1 entriesToSave)
slskd                   |    at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(StateManager stateManager, Boolean acceptAllChangesOnSuccess)
slskd                   |    at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.<>c.<SaveChanges>b__104_0(DbContext _, ValueTuple`2 t)
slskd                   |    at Microsoft.EntityFrameworkCore.Storage.NonRetryingExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
slskd                   |    at Microsoft.EntityFrameworkCore.DbContext.SaveChanges(Boolean acceptAllChangesOnSuccess)
```

I tried deleting the `transfers.db` files to no effect. I rolled back to `0.16.28` and it seems a bit better however adding a few dozen files from one user gave me a 500.

**Comment by jpdillingham** [comment]
The database can sometimes remain locked if the application doesn't exit cleanly, or it's possible a process has hung (#638), or for whatever reason two instances of the app managed to start.  Killing all instances of `slskd` or (worst case) rebooting the host should resolve it.

A 500 error enqueuing files is a separate issue that indicates there may have been a timeout, or you potentially hit a quota on the other user's client.  You'll typically see a reason along with the 500 message; do you recall what it was?

**Comment by tylerball** [comment]
I'm running slskd via docker-compose so I'm doubtful of multiple instances. I also tried `docker-compose down` before `rm`ing the transfers db just to be sure.

The 500s show up next to the download button as generic "contact the application owner" messages with a bunch of HTML. I'll try reupdating to `0.16.29` tonight and gather some screenshots.

**Comment by tylerball** [comment]
After a bit more playing around I think this is more related to #652, but on ZFS in my case. I’m running slskd via docker with its data directory on a mirrored ZFS dataset composed of two SSDs.

Switching to `cache.storage_mode: memory` instantly solved these issues and the app is running great! I have tons of memory on this server so this should work fine for me, but let me know if there’s any additional info you’d like from me.

**Comment by jpdillingham** [comment]
Thanks for reporting back, glad things are working better.  Unfortunately the databases for searches and transfer are still stored on disk and there's currently no way to change that.  There's not nearly as much I/O for either of these databases, though, so hopefully things will continue to work without issue.

I have looked into optimizing SQLite for BTRFS, and it seems there may be some tweaks that improve performance for ZFS as well, but I'm afraid that if I start changing things it may negatively impact users on different filesystems, so I'm waiting to get a feel for how widespread the issue is.

**Comment by tylerball** [comment]
Yeah I spoke a bit too soon, once I queued a large number of downloads I started seeing lock errors on the transfers DB. I'll send along my logs to that email.

**Comment by jpdillingham** [comment]
Thanks for sending the logs over.  Unfortunately it looks like the same incompatibility between slow filesystems, Entity Framework, and SQLite.

Unfortunately at this point all I can do is recommend against using slskd on a system with a copy-on-write filesystem such as BTRFS and ZFS, or over a network (NFS, samba etc).

I could put SQLite back into `MULTITHREADED` mode, but updates to transfer and search records need to be strictly ordered to work properly, and with this mode enabled writes that are serialized at the application level can be persisted out of order due to the delta in latency of SQLite updates.  Simply put; this would fix the locking issue but introduce incorrect behavior all over (as seen in #598)

I'm really not sure where to go from here, but I'll keep thinking about it.

The only relief I can offer right now is the addition of a 'volatile mode' that stores everything in memory.  It'll completely bypass these problems but you'll lose all search and transfer data when the application is shut down.  This is how it worked for a long time so maybe that's not too bad.

**Comment by tylerball** [comment]
Persistent transfers is a great feature so I personally wouldn't want to make that compromise. I moved my data directory to my non-ZFS boot drive and everything is great.

**Comment by jpdillingham** [comment]
Glad to hear! I went ahead and added 'volatile mode' anyway in #751 in case someone might find it useful.

**Comment by jpdillingham** [comment]
I added some guidance in a [System Requirements](https://github.com/slskd/slskd/blob/master/docs/system_requirements.md#devices-using-a-copy-on-write-filesystem-such-as-btrfs-or-zfs) doc that covers this.  I'll close this issue; feel free to re-open if anything else pops up.

**Comment by ImUrX** [comment]
I wanted to know the origin of the volatile mode and found this, I wanted to ask why not the usage of [Write-ahead logging on SQLite](https://wiki.tnonline.net/w/Blog/SQLite_Performance_on_Btrfs)?

**Comment by jpdillingham** [comment]
Entity Framework Core (the ORM I'm using) enables WAL automatically.

I was thinking recently about greatly reducing the number of database writes during transfers; right now it does so every time it updates every time the transfer progress is updated, which is potentially thousands of times per transfer.  I could throttle this and I think it would be fine, but the UI's updates would suffer.  I have an issue somewhere to change this over from polling (which hits the database) to streaming (which would have the app push data to the browser over a socket in real time).

So in summary, all of the options have been tweaked and benchmarked at this point; the path forward is just reducing database writes.
